sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala

```diff
@@ -35,8 +36,9 @@ import org.apache.spark.sql.types.StringType
  * This object targets to integrate various UDF test cases so that Scalar UDF, Python UDF and
  * Scalar Pandas UDFs can be tested in SBT & Maven tests.
  *
- * The available UDFs cast input to strings, which take one column as input and return a string
- * type column as output.
+ * The available UDFs are special. It defines an UDF wrapped by cast. So, Input column is casted
+ * into string, UDF returns strings as are, and then output column is casted back to the input
+ * column. In this way, UDF is virtually no-op.
```

Comment:
Yeah, I think it makes more sense.

Reply:
Good idea.

Reply:
I think virtually identical before / after. 

To clarify, complex types such as struct, array and map types cannot be roundtroup in string conversion - for complex tests let's workaround for those types.

Most of other types can be roundtroup. This will let us to avoid to use ugly workarounds for this case like `CAST` or `upper`.


Another one to note is that, since we should refer the input type to cast it back when we create expressions initially, it's required to use resolved expressions. Due to this, I had to add one restriction when it's used in Scala API (therefore it's unrelated when adding SQL tests at SPARK-27921).

See https://github.com/apache/spark/pull/25130/files#diff-893577587405a826f9c454b675073f75R983 . The input columns should always be resolved via `df.col(...)` or `df(...)` for now when Python or Pandas UDFs are used in Scala APIs by `IntegratedUDFTestUtils`.

Reply:
Not sure if we want support for complex types? If we just need this for query plan test in JVM for PythonUDFs, complex types make difference for the purpose? We can consider that later.